InvalidPathException on Windows with web-dependency-locator and webRoot

### Describe the bug

On Windows, with the quarkus-web-dependency-locator extension included and `src/main/resources/web/test.html` present, an InvalidPathException occurs. The exception indicates that `web/*` contains an illegal character (`*`).

The triggering `HotDeploymentWatchedFileBuildItem` is created here:
https://github.com/quarkusio/quarkus/blob/3.11.2/extensions/web-dependency-locator/deployment/src/main/java/io/quarkus/webdependency/locator/deployment/WebDependencyLocatorProcessor.java#L58

Though the error gets triggered here:
https://github.com/quarkusio/quarkus/blob/68b4cae73de2fd7eaccd1ba08e85e8ce20bb83d7/core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java#L1207

The `Paths.get()` will end in `sun.nio.fs.WindowsPathParser` rejecting the input because of that `*`.

### Expected behavior

Successful start.

### Actual behavior

An exception occurs:

```
Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.dev.HotDeploymentWatchedFileBuildStep#setupWatchedFileHotDeployment threw an exception: java.nio.file.InvalidPathException: Illegal char <*> at index 4: web/*
```

### How to Reproduce?

1. `quarkus.cmd create app -x quarkus-web-dependency-locator`
2. `cd .\code-with-quarkus\`
3. `mkdir .\src\main\resources\web`
4. `echo 'Test' > .\src\main\resources\web\test.html`
5. `quarkus.cmd dev`
    * Exception: `java.nio.file.InvalidPathException: Illegal char <*> at index 4: web/*`
6. `rm .\src\main\resources\web\test.html`
7. `rmdir .\src\main\resources\web\`
8. `quarkus.cmd dev`
    * Successful start

### Output of `uname -a` or `ver`

Windows 11 Pro (version 10.0.22631 N/A Build 22631)

### Output of `java -version`

openjdk version "21.0.2" 2024-01-16

### Quarkus version or git rev

3.11.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

Full stack trace:

```
2024-06-17 02:56:05,604 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-2) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.dev.HotDeploymentWatchedFileBuildStep#setupWatchedFileHotDeployment threw an exception: java.nio.file.InvalidPathException: Illegal char <*> at index 4: web/*
        at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:204)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:175)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
        at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
        at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:231)
        at java.base/java.nio.file.Path.of(Path.java:148)
        at java.base/java.nio.file.Paths.get(Paths.java:69)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.setWatchedFilePathsInternal(RuntimeUpdatesProcessor.java:1206)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.setWatchedFilePaths(RuntimeUpdatesProcessor.java:1146)
        at io.quarkus.deployment.dev.HotDeploymentWatchedFileBuildStep.setupWatchedFileHotDeployment(HotDeploymentWatchedFileBuildStep.java:37)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)

        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:331)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:265)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:58)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:191)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.restartCallback(IsolatedDevModeMain.java:174)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:547)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:447)
        at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$6.call(VertxHttpHotReplacementSetup.java:161)
        at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$6.call(VertxHttpHotReplacementSetup.java:148)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$0(ContextImpl.java:178)
        at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:279)
        at io.vertx.core.impl.ContextImpl.lambda$internalExecuteBlocking$2(ContextImpl.java:210)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1495)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.dev.HotDeploymentWatchedFileBuildStep#setupWatchedFileHotDeployment threw an exception: java.nio.file.InvalidPathException: Illegal char <*> at index 4: web/*
        at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:204)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:175)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
        at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
        at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:231)
        at java.base/java.nio.file.Path.of(Path.java:148)
        at java.base/java.nio.file.Paths.get(Paths.java:69)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.setWatchedFilePathsInternal(RuntimeUpdatesProcessor.java:1206)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.setWatchedFilePaths(RuntimeUpdatesProcessor.java:1146)
        at io.quarkus.deployment.dev.HotDeploymentWatchedFileBuildStep.setupWatchedFileHotDeployment(HotDeploymentWatchedFileBuildStep.java:37)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)

        at io.quarkus.builder.Execution.run(Execution.java:123)
        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:327)
        ... 19 more
Caused by: java.nio.file.InvalidPathException: Illegal char <*> at index 4: web/*
        at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:204)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:175)
        at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
        at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
        at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:231)
        at java.base/java.nio.file.Path.of(Path.java:148)
        at java.base/java.nio.file.Paths.get(Paths.java:69)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.setWatchedFilePathsInternal(RuntimeUpdatesProcessor.java:1206)
        at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.setWatchedFilePaths(RuntimeUpdatesProcessor.java:1146)
        at io.quarkus.deployment.dev.HotDeploymentWatchedFileBuildStep.setupWatchedFileHotDeployment(HotDeploymentWatchedFileBuildStep.java:37)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:483)
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java
@@ -1204,11 +1204,32 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
                     // Then process glob patterns
                     for (Entry<String, Boolean> e : watchedFilePaths.entrySet()) {
                         String watchedFilePath = e.getKey();
-                        Path path = Paths.get(watchedFilePath);
-                        if (!path.isAbsolute() && !watchedRootPaths.contains(e.getKey()) && maybeGlobPattern(watchedFilePath)) {
-                            Path resolvedPath = root.resolve(watchedFilePath);
-                            for (WatchedPath extra : expandGlobPattern(root, resolvedPath, watchedFilePath, e.getValue())) {
-                                timestamps.watchedPaths.put(extra.filePath, extra);
+                        Path path = Paths.get(sanitizedPattern(watchedFilePath));
+                        if (!path.isAbsolute() && !watchedRootPaths.contains(e.getKey())
+                                && maybeGlobPattern(watchedFilePath)) {
+                            try {
+                                final PathMatcher matcher = FileSystems.getDefault().getPathMatcher("glob:" + watchedFilePath);
+                                Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+                                    @Override
+                                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                                        Path relativePath = root.relativize(file);
+                                        if (matcher.matches(relativePath)) {
+                                            log.debugf("Glob pattern [%s] matched %s from %s", watchedFilePath, relativePath,
+                                                    root);
+                                            WatchedPath extra = new WatchedPath(file, relativePath, e.getValue(),
+                                                    attrs.lastModifiedTime().toMillis());
+                                            timestamps.watchedPaths.put(extra.filePath, extra);
+                                        }
+                                        return FileVisitResult.CONTINUE;
+                                    }
+
+                                    @Override
+                                    public FileVisitResult visitFileFailed(Path file, IOException exc) {
+                                        return FileVisitResult.CONTINUE;
+                                    }
+                                });
+                            } catch (IOException ex) {
+                                throw new UncheckedIOException(ex);
                             }
                         }
                     }
@@ -1219,8 +1240,9 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
         // Finally process watched absolute paths
         for (Entry<String, Boolean> e : watchedFilePaths.entrySet()) {
             String watchedFilePath = e.getKey();
-            Path path = Paths.get(watchedFilePath);
+            Path path = Paths.get(sanitizedPattern(watchedFilePath));
             if (path.isAbsolute()) {
+                path = Paths.get(watchedFilePath);
                 log.debugf("Watch %s", path);
                 if (Files.exists(path)) {
                     putLastModifiedTime(path, path, e.getValue(), timestamps);
@@ -1233,6 +1255,10 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
 
         log.debugf("Watched paths: %s", timestamps.watchedPaths.values());
         return this;
+    }
+
+    private String sanitizedPattern(String pattern) {
+        return pattern.replaceAll("[*?]", "");
     }
 
     private boolean maybeGlobPattern(String path) {
@@ -1280,32 +1306,6 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext, Closeable
         if (testClassChangeTimer != null) {
             testClassChangeTimer.cancel();
         }
-    }
-
-    private List<WatchedPath> expandGlobPattern(Path root, Path path, String pattern, boolean restart) {
-        PathMatcher pathMatcher = FileSystems.getDefault().getPathMatcher("glob:" + path.toString());
-        List<WatchedPath> files = new ArrayList<>();
-        try {
-            Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
-                @Override
-                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
-                    if (pathMatcher.matches(file)) {
-                        Path relativePath = root.relativize(file);
-                        log.debugf("Glob pattern [%s] matched %s from %s", pattern, relativePath, root);
-                        files.add(new WatchedPath(file, relativePath, restart, attrs.lastModifiedTime().toMillis()));
-                    }
-                    return FileVisitResult.CONTINUE;
-                }
-
-                @Override
-                public FileVisitResult visitFileFailed(Path file, IOException exc) {
-                    return FileVisitResult.CONTINUE;
-                }
-            });
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-        return files;
     }
 
     public boolean toggleInstrumentation() {

--- a/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/welcome/WelcomeProcessor.java
+++ b/extensions/vertx-http/deployment/src/main/java/io/quarkus/devui/deployment/welcome/WelcomeProcessor.java
@@ -72,11 +72,11 @@ public class WelcomeProcessor {
                     Path resourceDirs = resourcesDirs.iterator().next().getDir();
                     Path propertiesFile = resourceDirs.resolve("application.properties");
                     if (Files.exists(propertiesFile)) {
-                        return propertiesFile.toString().substring((int) root.length() + 1);
+                        return propertiesFile.toString().replace("\\", "/").substring((int) root.length() + 1);
                     }
                     Path propertiesYaml = resourceDirs.resolve("application.yaml");
                     if (Files.exists(propertiesYaml)) {
-                        return propertiesYaml.toString().substring((int) root.length() + 1);
+                        return propertiesYaml.toString().replace("\\", "/").substring((int) root.length() + 1);
                     }
                 }
             }
@@ -91,7 +91,7 @@ public class WelcomeProcessor {
                 String root = moduleDir.toPath().toString();
                 Collection<SourceDir> sourceDirs = workspaceModule.getMainSources().getSourceDirs();
                 if (sourceDirs != null && !sourceDirs.isEmpty()) {
-                    String sourceDir = sourceDirs.iterator().next().getDir().toString();
+                    String sourceDir = sourceDirs.iterator().next().getDir().toString().replace("\\", "/");
                     return sourceDir.substring((int) root.length() + 1);
                 }
             }
@@ -106,7 +106,7 @@ public class WelcomeProcessor {
                 String root = moduleDir.toPath().toString();
                 Collection<SourceDir> resourcesDirs = workspaceModule.getMainSources().getResourceDirs();
                 if (resourcesDirs != null && !resourcesDirs.isEmpty()) {
-                    String resourceDirs = resourcesDirs.iterator().next().getDir().toString();
+                    String resourceDirs = resourcesDirs.iterator().next().getDir().toString().replace("\\", "/");
                     return resourceDirs.substring((int) root.length() + 1);
                 }
             }

--- a/extensions/web-dependency-locator/deployment/src/main/java/io/quarkus/webdependency/locator/deployment/WebDependencyLocatorProcessor.java
+++ b/extensions/web-dependency-locator/deployment/src/main/java/io/quarkus/webdependency/locator/deployment/WebDependencyLocatorProcessor.java
@@ -55,7 +55,7 @@ public class WebDependencyLocatorProcessor {
                 .resolve(config.webRoot);
 
         if (Files.exists(web)) {
-            hotDeploymentWatchedProducer.produce(new HotDeploymentWatchedFileBuildItem(config.webRoot + SLASH + STAR));
+            hotDeploymentWatchedProducer.produce(new HotDeploymentWatchedFileBuildItem(config.webRoot + SLASH + STAR + STAR));
             // Find all css and js (under /app)
             Path app = web
                     .resolve(config.appRoot);
@@ -65,7 +65,8 @@ public class WebDependencyLocatorProcessor {
 
             if (Files.exists(app)) {
                 hotDeploymentWatchedProducer
-                        .produce(new HotDeploymentWatchedFileBuildItem(config.webRoot + SLASH + config.appRoot + SLASH + STAR));
+                        .produce(new HotDeploymentWatchedFileBuildItem(
+                                config.webRoot + SLASH + config.appRoot + SLASH + STAR + STAR));
                 try (Stream<Path> appstream = Files.walk(app)) {
                     appstream.forEach(path -> {
                         if (Files.isRegularFile(path) && path.toString().endsWith(DOT_CSS)) {
@@ -144,15 +145,16 @@ public class WebDependencyLocatorProcessor {
             try (StringWriter sw = new StringWriter()) {
                 if (!cssFiles.isEmpty()) {
                     for (Path css : cssFiles) {
-                        sw.write(TAB2 + "<link href='" + css.toString() + "' rel='stylesheet'>" + NL);
+                        sw.write(TAB2 + "<link href='" + css.toString().replace("\\", "/") + "' rel='stylesheet'>"
+                                + System.lineSeparator());
                     }
                 }
                 sw.write(TAB2 + IMPORTMAP_REPLACEMENT);
                 if (!jsFiles.isEmpty()) {
-                    sw.write(NL);
-                    sw.write(TAB2 + "<script type='module'>" + NL);
+                    sw.write(System.lineSeparator());
+                    sw.write(TAB2 + "<script type='module'>" + System.lineSeparator());
                     for (Path js : jsFiles) {
-                        sw.write(TAB2 + TAB + "import '" + js.toString() + "';" + NL);
+                        sw.write(TAB2 + TAB + "import '" + js.toString().replace("\\", "/") + "';" + System.lineSeparator());
                     }
                     sw.write(TAB2 + "</script>");
                 }
@@ -321,6 +323,4 @@ public class WebDependencyLocatorProcessor {
     private static final String DOT_HTML = ".html";
     private static final String DOT_CSS = ".css";
     private static final String DOT_JS = ".js";
-
-    private static final String NL = "\n";
 }

